src/binary/non_blocking/binary_buffer.rs

```diff
@@ -0,0 +1,856 @@
+use crate::binary::constants::v1_0::{length_codes, IVM};
+use crate::binary::int::DecodedInt;
+use crate::binary::non_blocking::type_descriptor::{
+    Header, TypeDescriptor, ION_1_0_TYPE_DESCRIPTORS,
+};
+use crate::binary::uint::DecodedUInt;
+use crate::binary::var_int::VarInt;
+use crate::binary::var_uint::VarUInt;
+use crate::result::{decoding_error, incomplete_data_error, incomplete_data_error_raw};
+use crate::types::integer::UInteger;
+use crate::{Integer, IonResult, IonType};
+use num_bigint::{BigInt, BigUint, Sign};
+use std::io::Read;
+use std::mem;
+
+// This limit is used for stack-allocating buffer space to encode/decode UInts.
+const UINT_STACK_BUFFER_SIZE: usize = 16;
+// This number was chosen somewhat arbitrarily and could be lifted if a use case demands it.
+const MAX_UINT_SIZE_IN_BYTES: usize = 2048;
+
+// This limit is used for stack-allocating buffer space to encode/decode Ints.
+const INT_STACK_BUFFER_SIZE: usize = 16;
+// This number was chosen somewhat arbitrarily and could be lifted if a use case demands it.
+const MAX_INT_SIZE_IN_BYTES: usize = 2048;
+
+/// A stack-allocated wrapper around an `AsRef<[u8]>` that provides methods to read Ion's
+/// encoding primitives.
+///
+/// When the wrapped type is a `Vec<u8>`, data can be appended to the buffer between read
+/// operations.
+#[derive(Debug, PartialEq)]
+pub(crate) struct BinaryBuffer<A: AsRef<[u8]>> {
```

Comment:
🗺️ This type allows you to read Ion encoding primitives from a `&[u8]`, `Vec<u8>`, `bytes::Buf`, or any other container of bytes.

There are many code blocks that are basically copy/pasted from the blocking implementation (for example: reading `VarUInt`, `VarInt`, etc). I intend to do a lot of consolidation in a follow-on PR.